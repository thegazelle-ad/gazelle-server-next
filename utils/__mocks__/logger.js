export const logger = {
  fatal: jest.fn().mockResolvedValue(undefined),
  error: jest.fn(),
  warn: jest.fn(),
  debug: jest.fn(),
};

logger.__mockClear = () => {
  logger.fatal.mockClear();
  logger.error.mockClear();
  logger.warn.mockClear();
  logger.debug.mockClear();
};
