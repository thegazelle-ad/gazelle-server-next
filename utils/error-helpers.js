const defaultError =
  'An error occured, the developers of The Gazelle have been notified';

/**
 * Builds a generic error message
 * @param {string} [msg] - The reason for the error
 * @returns {string}
 */
export const buildErrorMessage = msg => {
  if (!msg) {
    return defaultError;
  }
  return `${defaultError}\nReason: ${msg}`;
};
