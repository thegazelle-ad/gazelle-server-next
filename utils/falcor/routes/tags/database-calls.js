import _ from 'lodash';

import { database } from 'lib/db';

/**
 * @param   {Object[]}   tagObject the tag to be added.
 * @returns {boolean}    tags      number of tags modified
 */
export async function addTag(tagObject) {
  await database('tags').insert(tagObject);
  return true;
}

/**
 * Fetches direct meta data of tags from the tags database table
 * @param {string} queryField - Indicates which field to query by
 * @param {string[]} queryParams - Array of parameters of type queryField of articles to fetch
 * @param {string[]} columns - Which columns of the articles table to fetch
 * @returns {Promise<Object[]>}
 */
export function tagQuery(queryField, queryParams, columns) {
  // In order to be able to identify the rows we get back we need to include the queryField
  if (!columns.includes(queryField)) {
    columns.push(queryField);
  }
  return database
    .select(...columns)
    .from('tags')
    .whereIn(queryField, queryParams);
}

/**
 * Update tags given a json graph.
 * @param   {Object}   jsonGraphArg - the json graph to update
 */
export function updateTags(jsonGraphArg) {
  _.forEach(jsonGraphArg, async (tagObject, slug) => {
    await database('tags')
      .where('slug', '=', slug)
      .update(tagObject)
      .then(data => {
        if (data !== 1) {
          throw new Error(`Problems updating tag with slug: ${slug}`);
        }
      })
      .catch(e => {
        throw e;
      });
  });
}
