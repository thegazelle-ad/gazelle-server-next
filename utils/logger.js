/**
 * This file exposes our logger in a module form. It has the private variables
 * isClient and displayAlert that can only be accessed through the exposed API
 * of initializeLogger and updateDisplayAlert. If you come from the object
 * oriented programming paradigm this is equivalent to the singleton pattern,
 * except we never even created a class so we don't need to bother with making
 * sure it only gets instantiated once as it is now impossible.
 *
 * the logger object is the actual logger to be imported, but before it is used
 * initializeLogger should be called somewhere in the global scope, and if on the
 * client side and you want to add a prettier displayAlert function that is not necessarily
 * available at the start of the process, you can later call updateDisplayAlert
 * to make this your displayAlert function.
 */
import _ from 'lodash';

let isClient = null;
let displayAlert = null;

function validateInitialization() {
  if (isClient === null || (isClient && !_.isFunction(displayAlert)))
    throw new Error('The logger has yet to be initialized');
}

export const logger = {
  /**
   * Logs a fatal error. This should only be called if an error occurred that
   * completely breaks our system and requires a complete reboot of the code.
   * On the client it will refresh the page after displaying an alert,
   * on the server it will exit the server which will make 'forever' restart
   * the server.
   * @param {Error | string} err - The fatal error to be logged, either a string or
   * an Error object
   */
  async fatal(err) {
    validateInitialization();

    const errorObject = err instanceof Error ? err : new Error(err);
    if (isClient) {
      // We are actually lying currently, as developers haven't been notified
      // when we hook up sentry we will be notifying developers though
      await displayAlert(
        'A fatal error has occurred, and the developers' +
          ' have been notified. We apologize for the inconvenience, and we are afraid' +
          ' that we will have to reload the page now',
      );
      window.location.reload(true);
    } else {
      // eslint-disable-next-line no-console
      console.error(
        'A fatal error occured, and the server is exiting. See the error below:',
      );
      // eslint-disable-next-line no-console
      console.error(errorObject);
      process.exit(1);
    }
  },

  /**
   * Logs an error. This should only be called on unexpected behaviour that should
   * never happen, but it doesn't completely break our system
   * @param {Error | string} err - The error to be logged, either a string or
   * an Error object
   */
  error(err) {
    validateInitialization();

    const errorObject = err instanceof Error ? err : new Error(err);
    // eslint-disable-next-line no-console
    console.error(errorObject);
  },

  /**
   * Logs a warning. This should be called when an event occurs that is usually unwanted
   * but under certain circumstances could be okay
   * @param {Error | string} warning - The warning to be logged, either a string or
   * an Error object
   */
  warn(warning) {
    validateInitialization();

    // eslint-disable-next-line no-console
    console.warn(warning);
  },

  /**
   * Logs a debug statement. This can be called whereever the developer thinks
   * information could be helpful to have when debugging. It is the lowest logging
   * level so will normally only be chosen to be shown in development mode
   * @param {Error | string} msg - The warning to be logged, either a string or
   * an Error object
   */
  debug(msg) {
    validateInitialization();

    // eslint-disable-next-line no-console
    console.log(msg);
  },
};

/**
 * @param {boolean} updatedIsClient - Whether we are currently on the client or the server
 * @param {func} [updatedDisplayAlert] - This argument displays an alert on the client side
 * to the user, and should therefore only be supplied if isClient is true
 */
export const initializeLogger = (updatedIsClient, updatedDisplayAlert) => {
  if (updatedIsClient === undefined) {
    throw new Error('initializeLogger needs an isClient parameter supplied');
  }
  if (updatedIsClient && !_.isFunction(updatedDisplayAlert)) {
    throw new Error(
      'On the client initializeLogger needs a displayAlert function supplied',
    );
  }
  isClient = updatedIsClient;
  displayAlert = updatedDisplayAlert;
};

/**
 * Should only be called on the client, and should be used for updating the alert
 * function to a prettier one that is not initialized before initial render.
 * Expected use case is using window.alert as default and then switching it when
 * React has initialized and a nicer alert function has been created.
 * @param {func} newDisplayAlert - This argument displays an alert on the client side
 * to the user, and should therefore only be supplied if isClient is true
 */
export const updateDisplayAlert = newDisplayAlert => {
  if (!_.isFunction(newDisplayAlert)) {
    throw new Error(
      'updateDisplayAlert needs an alert function passed as a parameter',
    );
  }
  displayAlert = newDisplayAlert;
};
