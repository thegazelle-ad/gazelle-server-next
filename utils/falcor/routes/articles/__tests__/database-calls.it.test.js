import { getPaginatedArticle, createNewArticle } from '../database-calls.sql';
import {
  getDatabaseConnection,
  initializeTestDatabase,
  cleanupTestDatabase,
} from '__tests__/helpers/db-helpers';
import path from 'path';
import MockDate from 'mockdate';
import { buildErrorMessage } from 'lib/error-helpers';
import { logger } from 'lib/logger';

jest.mock('lib/logger');

const name = path
  .basename(__filename)
  .replace(/[.-]/g, '_')
  .replace(/^_+/, '');

const databaseName = `gazelle_test_${name}_${new Date().getTime()}`;

const database = getDatabaseConnection(databaseName);

const maxSlugNum = 200;

// End the knex connection after all tests are done
afterAll(async () => database.destroy());

describe('getPaginatedArticle', () => {
  // We can use All instead of Each as these are only select queries
  beforeAll(async () => initializeTestDatabase(database, databaseName));
  afterAll(async () => cleanupTestDatabase(databaseName));

  it('fetches expected rows from database', async () => {
    await expect(
      getPaginatedArticle(database, 5, 0),
    ).resolves.toMatchSnapshot();
  });

  it('fetches the newest article for page 0', async () => {
    await expect(getPaginatedArticle(database, 1, 0)).resolves.toEqual([
      { slug: `slug-${maxSlugNum}` },
    ]);
  });

  it('correctly computes offset', async () => {
    await expect(getPaginatedArticle(database, 1, 5)).resolves.toEqual([
      { slug: `slug-${maxSlugNum - 5}` },
    ]);
  });

  it('throws on negative page index', async () => {
    await expect(getPaginatedArticle(database, 1, -1)).rejects.toThrow();
  });

  it('throws on negative page length', async () => {
    await expect(getPaginatedArticle(database, -1, 5)).rejects.toThrow();
  });

  it('returns nothing on zero page length', async () => {
    await expect(getPaginatedArticle(database, 0, 5)).resolves.toEqual([]);
  });
});

describe('createNewArticle', () => {
  const insertMock = jest.fn().mockResolvedValue([1]);
  const databaseMock = jest.fn().mockReturnValue({ insert: insertMock });

  beforeAll(() => {
    // The number is irellevant as long as it is constant, this was the time the test
    // was written. It makes new Date() always return the same value so we can make
    // assertions about it (otherwise it would always return current time)
    MockDate.set(1524998880085);
  });

  afterAll(() => {
    MockDate.reset();
  });

  afterEach(() => {
    insertMock.mockClear();
    logger.__mockClear();
  });

  it('has expected input and output', async () => {
    const articleData = {
      slug: 'test-slug',
      title: 'test-title',
      teaser: 'test-teaser',
      imageUrl: 'https://google.com/img.png',
      category: 3,
      authors: [{ id: 1 }, { id: 2 }],
      tags: [{ id: 5 }, { id: 6 }],
    };
    const createdArticle = await createNewArticle(databaseMock, articleData);
    expect(createdArticle).toMatchSnapshot();
    expect(insertMock.mock.calls).toMatchSnapshot();
    articleData.authors.forEach((author, index) => {
      expect(author.id).toBe(createdArticle.authors[index].author_id);
    });
    articleData.tags.forEach((tag, index) => {
      expect(tag.id).toBe(createdArticle.tags[index].tag_id);
    });
  });

  it('errors on invalid author passed', async () => {
    const articleData = {
      authors: [{ slug: 'slug' }],
      slug: 'slug',
      title: 'title',
    };

    await expect(
      createNewArticle(databaseMock, articleData),
    ).rejects.toThrowError(buildErrorMessage());
    expect(logger.error.mock.calls).toMatchSnapshot();
  });

  it('errors on invalid tag passed', async () => {
    const articleData = {
      tags: [{ slug: 'slug' }],
      slug: 'slug',
      title: 'title',
    };

    await expect(
      createNewArticle(databaseMock, articleData),
    ).rejects.toThrowError(buildErrorMessage());
    expect(logger.error.mock.calls).toMatchSnapshot();
  });

  it('works with minimal input', async () => {
    // We just want to know it didn't error, so it's fine without any expect statements
    await createNewArticle(databaseMock, { slug: 'slug', title: 'title' });
  });

  it('throws without any slug', async () => {
    await expect(
      createNewArticle(databaseMock, { title: 'title' }),
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(logger.warn.mock.calls).toMatchSnapshot();
  });

  it('throws without any title', async () => {
    await expect(
      createNewArticle(databaseMock, { slug: 'slug' }),
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(logger.warn.mock.calls).toMatchSnapshot();
  });
});
