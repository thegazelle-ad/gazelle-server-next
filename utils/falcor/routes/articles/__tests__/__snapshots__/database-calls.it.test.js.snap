// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createNewArticle errors on invalid author passed 1`] = `
Array [
  Array [
    [Error: All authors in new article must have an id passed],
  ],
]
`;

exports[`createNewArticle errors on invalid tag passed 1`] = `
Array [
  Array [
    [Error: All tags in new article must have an id passed],
  ],
]
`;

exports[`createNewArticle has expected input and output 1`] = `
Object {
  "authors": Array [
    Object {
      "article_id": 1,
      "author_id": 1,
    },
    Object {
      "article_id": 1,
      "author_id": 2,
    },
  ],
  "category_id": 3,
  "created_at": 2018-04-29T10:48:00.085Z,
  "id": 1,
  "image_url": "https://google.com/img.png",
  "slug": "test-slug",
  "tags": Array [
    Object {
      "article_id": 1,
      "tag_id": 5,
    },
    Object {
      "article_id": 1,
      "tag_id": 6,
    },
  ],
  "teaser": "test-teaser",
  "title": "test-title",
}
`;

exports[`createNewArticle has expected input and output 2`] = `
Array [
  Array [
    Object {
      "category_id": 3,
      "created_at": 2018-04-29T10:48:00.085Z,
      "image_url": "https://google.com/img.png",
      "slug": "test-slug",
      "teaser": "test-teaser",
      "title": "test-title",
    },
  ],
  Array [
    Array [
      Object {
        "article_id": 1,
        "author_id": 1,
      },
      Object {
        "article_id": 1,
        "author_id": 2,
      },
    ],
  ],
  Array [
    Array [
      Object {
        "article_id": 1,
        "tag_id": 5,
      },
      Object {
        "article_id": 1,
        "tag_id": 6,
      },
    ],
  ],
]
`;

exports[`createNewArticle throws without any slug 1`] = `"Both slug and title must be provided for new articles"`;

exports[`createNewArticle throws without any slug 2`] = `
Array [
  Array [
    "Someone tried creating an article without slug or title",
  ],
]
`;

exports[`createNewArticle throws without any title 1`] = `"Both slug and title must be provided for new articles"`;

exports[`createNewArticle throws without any title 2`] = `
Array [
  Array [
    "Someone tried creating an article without slug or title",
  ],
]
`;

exports[`getPaginatedArticle fetches expected rows from database 1`] = `
Array [
  RowDataPacket {
    "slug": "slug-200",
  },
  RowDataPacket {
    "slug": "slug-199",
  },
  RowDataPacket {
    "slug": "slug-198",
  },
  RowDataPacket {
    "slug": "slug-197",
  },
  RowDataPacket {
    "slug": "slug-196",
  },
]
`;
